### PR TITLE
Removed DenyEscalatingExec from the list of default admission control…

### PR DESCRIPTION
**Pull request by ArtfulCoder**
#14934

@bgrant0607

### CODE REVIEW

**Comment by ArtfulCoder** [comment]
#14934

**Comment by bgrant0607** [comment]
LGTM. Thanks. Please cherrypick this after merge.

**Comment by ArtfulCoder** [comment]
@k8s-bot test this please

**Comment by liggitt** [comment]
cc @pweil-

**Comment by derekwaynecarr** [comment]
Hmm... I guess there must be a use case you have in mind for wanting to disable this?  

Is there an example or something this broke by having it enabled?

**Comment by ArtfulCoder** [comment]
It is an inconvenience to users and does not provide any real security.

**Comment by a-robinson** [comment]
Needs a rebase before merge

**Comment by ArtfulCoder** [comment]
rebased

File: hack/local-up-cluster.sh [hunk 1]
```diff
@@ -202,11 +201,7 @@ function set_service_accounts {
 
 function start_apiserver {
     # Admission Controllers to invoke prior to persisting objects in cluster
-    if [[ -z "${ALLOW_SECURITY_CONTEXT}" ]]; then
-      ADMISSION_CONTROL=NamespaceLifecycle,NamespaceAutoProvision,LimitRanger,SecurityContextDeny,ServiceAccount,DenyEscalatingExec,ResourceQuota
-    else
-      ADMISSION_CONTROL=NamespaceLifecycle,NamespaceAutoProvision,LimitRanger,ServiceAccount,DenyEscalatingExec,ResourceQuota
-    fi
+    ADMISSION_CONTROL=NamespaceLifecycle,NamespaceAutoProvision,LimitRanger,SecurityContextDeny,ServiceAccount,ResourceQuota
```
Comment (a-robinson) [inline hack/local-up-cluster.sh]: This looks like a bad rebase - you appear to have clobbered the change that you were supposed to rebase on top of

**Comment by a-robinson** [comment]
Removing lgtm for now to avoid auto-merge

File: hack/local-up-cluster.sh (hunk 1 above)
Comment (ArtfulCoder) [inline hack/local-up-cluster.sh]: I don't think so.
I want it to be this way

**Comment by ArtfulCoder** [comment]
@a-robinson I think I merged it the way I wanted it to be. I deliberately removed the if check since its not required with my change.

File: hack/local-up-cluster.sh (hunk 1 above)
Comment (a-robinson) [inline hack/local-up-cluster.sh]: So the ALLOW_SECURITY_CONTEXT parameter isn't required anymore? It was only added this morning in #15081

Comment (a-robinson) [inline hack/local-up-cluster.sh]: sorry - s/required/supported

Comment (pmorie) [inline hack/local-up-cluster.sh]: @ArtfulCoder the if statement removed in the current diff needs to be restored.  We need a way to enable security context on local clusters.

**Comment by ArtfulCoder** [comment]
@bgrant0607 for LGTM post-rebase

**Comment by liggitt** [comment]
@ArtfulCoder your change wasn't related to the SecurityContextDeny admission controller, was it?

**Comment by liggitt** [comment]
The combination of this PR and https://github.com/kubernetes/kubernetes/pull/15021 means that the defaults now 1) allow privileged pods from the API and 2) allow users to exec into privileged pods. That's a very impactful change, security-wise.

That should be publicized on the mailing list, release notes, etc... along with instructions to get back to the previous defaults for people who don't want their users to be able to take over their nodes

**Comment by pmorie** [comment]
@ArtfulCoder 

> I think I merged it the way I wanted it to be. I deliberately removed the if check since its not required with my change.

It is still required with your change -- the state `local-up-cluster.sh` is in in your current diff leaves `SecurityContextDeny` in place.  Please keep the changes from #15081, we still need them.

> That should be publicized on the mailing list, release notes, etc... along with instructions to get back to the previous defaults for people who don't want their users to be able to take over their nodes

+1 to @liggitt's suggestion

**Comment by bgrant0607** [comment]
This is not at all impactful security-wise. 

```
apiVersion: v1
kind: Pod
metadata:
  name: priv-pod
spec:
  volumes:
    - name: docker-sock
      hostPath:
        path: /run/docker.sock
  containers:
    - name: unpriv
      image: docker
      command:
        - /bin/sh
        - -c
        - "docker run --privileged busybox /bin/sh -c 'echo success'"
      volumeMounts:
        - name: docker-sock
          mountPath: /var/run/docker.sock
```

File: hack/local-up-cluster.sh (hunk 1 above)
Comment (bgrant0607) [inline hack/local-up-cluster.sh]: We need a way to remove SecurityContextDeny. I'd be in favor of omitting it by default, in the same vein as enabling everything else by default in the absence of real, comprehensive security policy.

**Comment by liggitt** [comment]
1. I set up a cluster running two types of images:
  * privileged infrastructure images
  * non-privileged, well-known images

2. I grant users access to exec/attach to pods in order to use the non-privileged, well-known images I am running, but I prevent them from creating their own pods/rcs, and the DenyEscalatingExec admission controller keeps them from running commands inside one of my privileged infrastructure images.

3. I restart with the latest scripts, and the DenyEscalatingExec protection is no longer there, and my infrastructure pods are exposed.

File: hack/local-up-cluster.sh (hunk 1 above)
Comment (pmorie) [inline hack/local-up-cluster.sh]: i would also be okay with omitting it by default.  I'm writing E2Es around
it now and i would like to be able to them run in the google jenkins.

**Comment by liggitt** [comment]
Again, just saying that we should publicize this change, so that anyone who has a use case like that knows they need to take steps themselves

**Comment by bgrant0607** [comment]
Motivation for enabling all functionality by default: Disabling features like privileged, hostNetworking, securityContext, etc. just inconvenience users, for non-obvious (to users) reasons. We did not clearly document that these features were blocked by default (e.g., API documentation provides no hints about this) and have received a number of requests to enable them (e.g., privileged mode for NFS). This of this as a bug fix or enabling experimental features. It's purely additive and doesn't break compatibility. Since the default set of admission controllers didn't block access to /run/docker.sock, this doesn't change what the user could and could not do, only what was convenient to do.

**Comment by bgrant0607** [comment]
I agree that we do need to have the means by which users of the OSS Kubernetes can disable these features, which is why these admission controllers exist and are documented in the cluster admin guide.

**Comment by bgrant0607** [comment]
I also agree with documenting the change in the release notes.

**Comment by bgrant0607** [comment]
@liggitt I don't understand the example you described in https://github.com/kubernetes/kubernetes/pull/15035#issuecomment-145726943

You use the term "image" rather than "container" or "pod" or "namespace".

What's to stop a user from launching a pod on a specified node and doing `docker exec` to any container on the same node, or indeed just launching another privileged container?

**Comment by liggitt** [comment]
> What's to stop a user from launching a pod on a specified node and doing docker exec to any container on the same node, or indeed just launching another privileged container?

a policy that allows one set of users to create rcs/pods, and another set of users to attach/exec to them

**Comment by bgrant0607** [comment]
So the use case is someone using the default list of admission controllers but setting a custom authorization policy. Is this possible? I don't see a way to set --authorization-mode and --authorization-policy-file in cluster/saltbase/salt/kube-apiserver/kube-apiserver.manifest.

**Comment by liggitt** [comment]
I've lost track of all the ways the apiserver can be started/installed, so I'm not sure, but never underestimate what a desperate deployer can do :) I've seen scripted deployments use sed to manipulate the unit files directly to add/modify arguments.

Not saying we shouldn't make this change, just that a heads up is nice for people who are doing things we might not expect in combination with in-repo setup, and would be bitten if this changes under them.

**Comment by bgrant0607** [comment]
Totally agree we should announce the change.

**Comment by ArtfulCoder** [comment]
@bgrant0607 I re-pushed. PTAL

**Comment by bgrant0607** [comment]
LGTM

**Comment by pmorie** [comment]
Lgtm, thanks for fixing.

**Pull request merged**